### PR TITLE
Use rapids-build-backend.

### PR DESCRIPTION
**Pull request by bdice**
## Description
This PR uses `rapids-build-backend` to simplify wheel builds and reduce the complexity of various CI/build scripts.

See also:
- https://github.com/rapidsai/rapids-build-backend
- https://github.com/rapidsai/build-planning/issues/31
- https://github.com/rapidsai/cudf/pull/15245
- #1529

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rmm/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/pyproject.toml
Comment (KyleFromNVIDIA) [inline python/pyproject.toml:125]: With the [new commit handling](https://github.com/rapidsai/rapids-build-backend/pull/3), this should be:

```suggestion
commit-file = "rmm/GIT_COMMIT"
```

File: ci/build_wheel.sh
Comment (vyasr) [inline ci/build_wheel.sh:17]: I think before moving forward we should apply your suggestion from our previous meeting:
- Change the config option to `allow-nightly-deps`.
- Specify `allow-nightly-deps = 1` in the pyproject.toml in this PR
- Override that with the new environment variable `RAPIDS_ALLOW_NIGHTLY_DEPS=0` if `rapids-is-release-build` in this script.

File: python/rmm/pyproject.toml
Comment (bdice) [inline python/rmm/pyproject.toml:130]: At some point I'd like to consider switching the "default" pyproject.toml to CUDA 12. That can be evaluated later but if we like the idea we could file a build-planning issue.

**Review by bdice — COMMENTED** [review]
@jameslamb Thanks for picking this up! Since I filed the PR, I cannot approve it. However, I _would_ approve if I could. :) Feel free to self-approve this on my behalf since it's your work now.

File: dependencies.yaml
Comment (KyleFromNVIDIA) [inline dependencies.yaml:97]: As discussed at https://github.com/rapidsai/cuml/pull/5804#discussion_r1600244946, I think we should have an upper bound here (and all other instances as well):

```suggestion
          - rapids-build-backend >=0.1.1,<0.2.0.dev0
```

**Review by KyleFromNVIDIA — APPROVED** [review]
Approved with one minor suggestion

File: dependencies.yaml
Comment (jameslamb) [reply dependencies.yaml]: yep thanks! Pushed this in https://github.com/rapidsai/rmm/pull/1502/commits/a507ae1e4bfb81e596ef77cca2e26aa5c1f6daf9.

I went with `~=0.1.1` for the conda one, just out of minor personal preference. It'll achieve the same effect.

Comment (jameslamb) [reply dependencies.yaml]: blegh forgot the `rmm` conda recipe, added in https://github.com/rapidsai/rmm/pull/1502/commits/dcba4c2d696cf46b8a3d2d073475299df2b3d952

**Comment by jameslamb** [comment]
We're not targeting the 24.06 for this. I'll keep the `DO NOT MERGE` label on and retarget this at `branch-24.08` once that exists.

**Comment by jameslamb** [comment]
Since I retargeted this at `branch-24.08`, the two `devcontainer` builds are failing like this:

```text
  About to run devcontainer build --workspace-folder /home/runner/_work/rmm/rmm --config /home/runner/_work/rmm/rmm/.devcontainer/cuda12.2-conda/devcontainer.json --image-name undefined:latest
  [2024-05-21T14:56:50.141Z] @devcontainers/cli 0.60.0. Node.js v16.20.2. linux 6.5.0-1020-aws x64.
  
  Error: Dev container build failed: Command failed: docker inspect --type image rapidsai/devcontainers:24.08-cpp-mambaforge-ubuntu22.04 (exit code: undefined)
  An error occurred building the container.
  Error: Command failed: docker inspect --type image rapidsai/devcontainers:24.08-cpp-mambaforge-ubuntu22.04
```

([build link](https://github.com/rapidsai/rmm/actions/runs/9176786690/job/25232887811?pr=1502))

That `--image-name undefined:latest` looks suspicious, but I'd have to dig a bit to find where that's coming from. Saw something similar this morning on a `cuml` PR targeting `branch-24.08`: https://github.com/rapidsai/cuml/pull/5901

**Comment by vyasr** [comment]
The devcontainers probably hadn't been built yet, those builds were manually kicked off just 4 hours ago. I've kicked off a rerun to test.

**Comment by jameslamb** [comment]
Yep that worked ([build link](https://github.com/rapidsai/rmm/actions/runs/9176786690/job/25248301694?pr=1502)), thanks @vyasr

File: ci/build_python.sh
Comment (vyasr) [inline ci/build_python.sh:16]: Out of scope for this PR, but a small idea for improvement: we ought to change `rapids-generate-version` to something like `rapids-overwrite-version <path_to_file>` that also does the echo overwriting. Then these build scripts become nearly trivial.

Even more out of scope for this PR: I would probably recommend tackling https://github.com/rapidsai/build-planning/issues/49 before making any further changes to gha-tools.

File: dependencies.yaml
Comment (vyasr) [inline dependencies.yaml:171]: Needs an alpha spec too. Not critical for correctness here, but I think https://github.com/rapidsai/pre-commit-hooks/pull/28 will enforce this now.

**Review by vyasr — APPROVED** [review]
Couple of small things, but approving since I don't need to look at this again unless something comes up.

**Comment by KyleFromNVIDIA** [comment]
> do we also need to add a disable-cuda in the build.sh invocations?

Yes, but it needs to be `rapidsai.disable-cuda` once https://github.com/rapidsai/rapids-build-backend/pull/37 is merged.

File: ci/build_python.sh
Comment (jameslamb) [reply ci/build_python.sh:16]: We could get that behavior (cutting out an assignment here) with 0 changes to `rapids-generate-version`.

This:

```bash
version=$(rapids-generate-version)
echo "${version}" > VERSION
# ...
RAPIDS_PACKAGE_VERSION=${version} rapids-conda-retry mambabuild -c "${CPP_CHANNEL}" conda/recipes/rmm
```

Could become this:

```bash
rapids-generate-version > VERSION
# ...
RAPIDS_PACKAGE_VERSION=${head -1 VERSION} rapids-conda-retry mambabuild -c "${CPP_CHANNEL}" conda/recipes/rmm
```

I'm gonna need to push another commit here anyway to add the alpha spec and maybe the config setting for `disable-cuda` (once https://github.com/rapidsai/rapids-build-backend/pull/37 is released). Want me to make that change too?

Comment (vyasr) [reply ci/build_python.sh:16]: Yeah sure, why not. I'm all for brevity. Let's port that to all the repos while we're at it (CC @KyleFromNVIDIA).

Comment (jameslamb) [reply ci/build_python.sh]: Alright thanks, I just pushed commits here doing that for all uses of `rapids-generate-version` in these build scripts.

**Comment by jameslamb** [comment]
Alright I've updated this with the following changes:

* moved `rapids-build-backend` floor to 0.3.0
* passed `--config-settings rapidsai.disable-cude=true` in `build.sh`
* consolidated `rapids-generate-version` calls based on https://github.com/rapidsai/rmm/pull/1502#discussion_r1611757961
* changed `librmm/_version.py` to also use the `__package__` pattern (instead of string literal `"librmm"`)
* added `__all__` lists to `_version.py`
   - *(noticed @KyleFromNVIDIA doing this in https://github.com/rapidsai/cuml/pull/5804, I think it's nice to clarify intent and to avoid re-exporting other things from that module)*

That's enough changes that I think this should get another review if you have time @vyasr .

File: ci/build_python.sh
Comment (KyleFromNVIDIA) [reply ci/build_python.sh]: I've done the same in my PRs.

File: ci/build_cpp.sh
Comment (vyasr) [inline ci/build_cpp.sh:16]: In the conda case where we need the value again to set the env var for the conda build I don't think piping then reading back from file is an improvement over the original approach. I won't block merging over that though, up to you if you like this better or if you don't but don't think it's worth another iteration.

File: dependencies.yaml
Comment (vyasr) [inline dependencies.yaml:97]: You upper bounded in the pyproject section but not here. We should do so consistently (CC @KyleFromNVIDIA)

**Review by vyasr — APPROVED** [review]
Couple more small notes, but approving assuming that they're addressed.

File: dependencies.yaml
Comment (bdice) [reply dependencies.yaml:97]: Does conda even know how to use `~=`? I thought only pip used that syntax.

Comment (jakirkham) [reply dependencies.yaml:97]: In recipes, but not in environment files. The recipe syntax actually just gets converted to comparison operators in the final package metadata

Comment (jameslamb) [reply dependencies.yaml:97]: > Does conda even know how to use `~=`

Yes: https://docs.conda.io/projects/conda-build/en/latest/resources/package-spec.html#package-match-specifications.

But anyway I think this was being too cute for no real benefit. I'll change the conda ones to `>=x.x.x,<x.x.x`, matching the `pyproject` and `requirements` ones.

File: ci/build_cpp.sh
Comment (jameslamb) [reply ci/build_cpp.sh:16]: Ha I was anticipating a review comment of the form "what about `build_cpp.sh`, that's now inconsistent with the others" if it wasn't changed this way.

I'm gonna push another commit anyway in response to https://github.com/rapidsai/rmm/pull/1502#discussion_r1612462557, so I'll revert this.

Comment (jameslamb) [reply ci/build_cpp.sh]: reverted in https://github.com/rapidsai/rmm/pull/1502/commits/7aeba946df9a5c50f513e9516afb33c27a5c0435.

File: dependencies.yaml
Comment (jameslamb) [reply dependencies.yaml]: changed these all to `>=0.3.0,<0.4.0.dev0` in https://github.com/rapidsai/rmm/pull/1502/commits/7aeba946df9a5c50f513e9516afb33c27a5c0435

I'll do the same in https://github.com/rapidsai/cudf/pull/15245

**Comment by trivialfis** [comment]
I think the README file needs to be updated.

**Comment by jakirkham** [comment]
Could you please raise this in an issue or PR?

**Comment by trivialfis** [comment]
@jakirkham Opened an issue here: https://github.com/rapidsai/rmm/issues/1609 .

**Pull request merged**
